### PR TITLE
[ CI/Build ] LM Eval Harness Based CI Testing

### LINKED ISSUE #5800: [Feature]: Integration Testing with lm-eval-harness
### 🚀 The feature, motivation and pitch

As vLLM added more capable nodes for CI, I think it's a good time to start adding model quality test for both non-quantized and quantized model to ensure the kernel and scheduler changes do not degrade the model accuracy performance. This also ensure vLLM doesn't break lm-eval-harness's integration. 

I would like to ask for suggestions for concrete benchmarks to be added. For example, MMLU for Llama3-8B with a score >= X. 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
SUMMARY:
* Adds lm-eval harness based testing, using GSM-8k
* Scripts to create baseline scores offline
* Scripts to run lm-eval integration tests
* Configs for large and small models

USAGE:
```bash
pip install git+https://github.com/EleutherAI/lm-evaluation-harness.git@262f879a06aa5de869e5dd951d0ff2cf2f9ba380
pip install openai==1.3.9
bash .buildkite/lm-eval-harness/run-lm-eval-vllm.sh -c .buildkite/lm-eval-harness/configs/small-models.txt 
```

FOLLOW UP:
* Same flow via `LLM`

FIX #5800

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by robertgshaw2-neuralmagic** [comment]
@simon-mo ready for review

I was unable to remove the second layer of indirection.

The reason for this is that, when TP>1, we cannot launch vllm more than once within the same process. This same issue hampers our existing distributed tests.

Alternatively, I could remove the `run-tests.sh` file and instead have that logic in the buildkite `test-pipeline.yaml`.

File: .buildkite/lm-eval-harness/test_lm_eval_correctness.py
Comment (simon-mo) [inline .buildkite/lm-eval-harness/test_lm_eval_correctness.py:30]: i'm not super familiar but is this using `LLM` or completion server?

**Review by simon-mo — APPROVED** [review]
LGTM once the A100 test passes. Also we should also confirm how long it takes to run (ideally quickly since we only have one box atm)

File: .buildkite/lm-eval-harness/test_lm_eval_correctness.py
Comment (robertgshaw2-neuralmagic) [reply .buildkite/lm-eval-harness/test_lm_eval_correctness.py:30]: - `LLM`, I removed the server so we could test the official integration rather than `local-completions`

**Comment by robertgshaw2-neuralmagic** [comment]
> LGTM once the A100 test passes. Also we should also confirm how long it takes to run (ideally quickly since we only have one box atm)

I limited it to 250 samples. Once the models are loaded its <1 min

**Comment by DarkLight1337** [comment]
Running A100 tests... please check the results before merging.

**Pull request merged**
